### PR TITLE
feat(ios): add Pushed Authorization Request (PAR) support

### PR DESCRIPTION
**Pull request by Kaushikgupta469**
### Implementation of PAR wrt to Specification
- AuthorizationServerMetadata: pushed_authorization_request_endpoint + require_pushed_authorization_requests
- PushedAuthorizationResponse model + PushedAuthorizationRequestException
- PushedAuthorizationRequestService (POST params -> request_uri)
- AuthorizationUrlBuilder.buildWithRequestUri (short client_id+request_uri URL)
- ImplicitAuthorizationRequestData carries PAR endpoint / authorization_details / issuer_state
- RedirectToWebAuthorizationMethodService branches to PAR when advertised, else falls back to the long-URL flow
- thread issuer_state + PAR endpoint through AuthorizationCodeFlowService
- unit tests for PushedAuthorizationRequestService

## Summary by CodeRabbit

* **New Features**
  * Added support for Pushed Authorization Requests (PAR), including a new PAR push capability.
  * Enhanced redirect-to-web authorization to optionally use PAR and build authorization links using `request_uri`.
  * Extended authorization server metadata with PAR endpoint and “require PAR” support.
* **Bug Fixes**
  * Improved PAR/redirect-to-web error handling to preserve issuer error details when available.
* **Tests**
  * Added PAR service tests for request construction, response parsing, and error propagation (including DPoP/PKCE/client-auth rules).
  * Updated CI to run tests on an iOS 17.5 simulator.

### CODE REVIEW

**Comment by Kaushikgupta469** [comment]
@coderabbitai Review this PR against the requirements below. This is the Swift/iOS port of PAR, mirroring the Kotlin inji-vci-client implementation (uses Codable, async/await).

Requirements:

Implement PAR in the Authorization Code Flow.
Discover the PAR endpoint from AS metadata.
Send auth request params (response_type, client_id, redirect_uri, code_challenge, code_challenge_method, authorization_details/scope, issuer_state if present) to the PAR endpoint.
Handle successful responses (request_uri, expires_in).
Redirect to the Authorization Endpoint using client_id + request_uri.
Error handling for: missing PAR endpoint, PAR request failures, invalid PAR responses, client auth failures.
Comprehensive test coverage for all the above.
Please review: requirement coverage, missing edge cases, security / OAuth / OID4VCI / RFC 9126 deviations, code quality, test gaps, and any bugs or regressions.

Swift-specific: confirm the explicit init on AuthorizationServerMetadata doesn't break synthesized Codable init(from:) (absent keys → nil); existing call sites still compile; PushedAuthorizationRequestException is caught/wrapped correctly in async paths; and request_uri / scope-vs-authorization_details handling is correct.

Review all changed files and give concrete suggestions with code examples.

File: Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift [hunk 1]
```diff
@@ -5,4 +5,25 @@ struct ImplicitAuthorizationRequestData: AuthorizationRequestData {
     let clientMetadata: ClientMetadata
     let pkceSession: PKCESessionManager.PKCESession
     let scope: String
+    let pushedAuthorizationRequestEndpoint: String?
+    let authorizationDetails: String?
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift]: `authorizationDetails` is never populated — no caller sets it on `ImplicitAuthorizationRequestData`, so the PAR service always falls through to `scope`. The authorization-code flow here is scope-based and doesn't support `authorization_details` (only the separate IAE flow does, via `IARInitialRequestBody`). Please drop this field along with the matching parameter and `if/else` branch in `PushedAuthorizationRequestService` and the pass-through in `RedirectToWebAuthorizationMethodService`, so the PR doesn't imply support that isn't wired.

File: Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift [hunk 2]
```diff
@@ -133,6 +133,8 @@ class AuthorizationCodeFlowService {
         do {
             let pkceSession = pkceSessionManager.createSession()
 
+            let issuerState = credentialOffer?.grants?.authorizationCodeGrant?.issuerState
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift]: `issuer_state` is only forwarded on the PAR branch; the non-PAR path (`AuthorizationUrlBuilder.build`) has no `issuer_state` parameter, so it's dropped there. That half-wires a conditional MUST (§4.1.1: if the offer carries `issuer_state`, the Wallet MUST include it in the Authorization Request). Forwarding it was already missing on develop, so let's track it as its own issue and remove the partial `issuerState` plumbing from this PAR PR (this line, the method params, and the pass-throughs) to keep the PR focused.

File: Sources/VCIClient/authorizationServer/AuthorizationUrlBuilder.swift [hunk 3]
```diff
@@ -25,6 +25,17 @@ enum AuthorizationUrlBuilder {
         return url
     }
 
+    static func buildWithRequestUri(
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationServer/AuthorizationUrlBuilder.swift:30]: Naming: keep this as a separate method rather than overloading `build` — the two produce semantically different requests (full params vs a `request_uri` reference), so overloading would be a footgun. `buildWithRequestUri` is clear as-is. If you want symmetry, optionally rename `build` → `buildWithParameters` in a follow-up, but don't overload.

**Review by abhip2565 — CHANGES_REQUESTED** [review]
Faithful PAR port — parity with the Kotlin PR is good. Same cleanups apply here: remove the unused `authorization_details` plumbing (never populated, flow is scope-based) and the partial `issuer_state` wiring (only on the PAR path — to be handled as a separate issue), plus a note on the builder method naming. Details inline.

**Comment by Kaushikgupta469** [comment]
@coderabbitai  Holding off the above reviews  as reason :

authorization_details (service + test): Intentionally removed in this PR — this flow is scope-based, and authorization_details is only wired through the separate IAR flow. Adding it back would imply support that isn't threaded end-to-end; the test asserting it's never sent is deliberate.

require-PAR flag: When a PAR endpoint is present the flow already fails closed (no fallback). Enforcing the require-flag is a known gap (also on Kotlin) — tracking separately to keep this PR focused.

expires_in: The request_uri is used immediately, so it isn't load-bearing, and the Kotlin lib this mirrors doesn't validate it either. Skipping for parity.

File: .github/workflows/push-trigger.yml [hunk 4]
```diff
@@ -32,7 +32,7 @@ jobs:
         run: xcodebuild -scheme VCIClient -destination 'generic/platform=iOS' build
 
       - name: Run tests
-        run: xcodebuild test -scheme VCIClientTests -destination 'platform=iOS Simulator,OS=latest,name=iPhone 15'
+        run: xcodebuild test -scheme VCIClientTests -destination 'platform=iOS Simulator,OS=17.5,name=iPhone 15'
```
Comment (abhip2565) [inline .github/workflows/push-trigger.yml:35]: all versions might not be present in Mac runner. Is latest causing some issues?

File: Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift [hunk 5]
```diff
@@ -371,14 +371,24 @@ class AuthorizationCodeFlowService {
...
+                tokenEndpointAuthMethodsSupported: authorizationServerMetadata.tokenEndpointAuthMethodsSupported,
+                requirePushedAuthorizationRequests: authorizationServerMetadata.requirePushedAuthorizationRequests
             )
 
             let response: AuthorizationResponse
             do {
                 response = try await RedirectToWebAuthorizationMethodService(
                     openWebPage: openWebPage
                 ).authorizeUser(requestData: requestData)
+            } catch let e as VCIClientException {
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift:383]: is this catch needed?

File: Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift [hunk 6]
```diff
@@ -5,5 +5,28 @@ struct ImplicitAuthorizationRequestData: AuthorizationRequestData {
     let clientMetadata: ClientMetadata
     let pkceSession: PKCESessionManager.PKCESession
     let scope: String
+    let pushedAuthorizationRequestEndpoint: String?
     let dpopJkt: String
+    let tokenEndpointAuthMethodsSupported: [String]?
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift]: Is this needed to parse?

File: Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift [hunk 7]
```diff
@@ -5,5 +5,28 @@ struct ImplicitAuthorizationRequestData: AuthorizationRequestData {
...
+    let requirePushedAuthorizationRequests: Bool?
+
+    init(
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift:11]: @coderabbitai is this init requred?

File: Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift [hunk 8]
```diff
@@ -20,16 +25,23 @@ final class RedirectToWebAuthorizationMethodService: AuthorizationMethodService
...
+        let parEndpoint = request.pushedAuthorizationRequestEndpoint
+        let isParRequired = request.requirePushedAuthorizationRequests
+
+        let authUrl: String
+        if let parEndpoint, !parEndpoint.isEmpty, isParRequired == true {
+            authUrl = try await buildAuthorizationUrlViaPushedRequest(
+                request: request,
+                parEndpoint: parEndpoint
+            )
+        } else if let parEndpoint, !parEndpoint.isEmpty, isParRequired == nil {
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift]: What if par if PAR required is false?

File: Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift [hunk 9]
```diff
@@ -20,16 +25,23 @@ final class RedirectToWebAuthorizationMethodService: AuthorizationMethodService
...
+            authUrl = try await buildAuthorizationUrlViaPushedRequest(
```
Comment (abhip2565) [inline Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift:32]: we should fallback to normal auth in that case right?

**Review by abhip2565 — CHANGES_REQUESTED** [review]

File: Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift (hunk 7 above)
Comment (abhip2565) [reply Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift:11]: where all is this called?

File: Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift (hunk 9 above)
Comment (Kaushikgupta469) [reply Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift:32]: yes for par requried = false it will be a fallback but as per spec if enpoint is there its alone is sufficient as disccued will update the logic then par will being hit if par enpoint url is present fallback will ony be in cas when no par url is there

File: Sources/VCIClient/authorizationServer/PushedAuthorizationRequestService.swift [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+            params["scope"] = scope
+        }
+
+        if let dpopJkt = dpopJkt, !dpopJkt.isEmpty {
+            params["dpop_jkt"] = dpopJkt
+        }
+
+        // Core authorization request params take precedence; clientAuthParams
+        // (e.g. client_assertion) may only add keys, never override reserved fields.
+        params.merge(clientAuthParams) { current, _ in current }
```
[the remark this answers is not in the document]
Comment (Kaushikgupta469) [reply Sources/VCIClient/authorizationServer/PushedAuthorizationRequestService.swift:40]: @coderabbitai
Correct that the merge only guards keys already present, so `scope`/`dpop_jkt` could be injected 
when they're nil/empty. But `clientAuthParams` isn't populated by any caller in either SDK today — 
it's an unused extension point — so there's no reachable defect.

Also `issuer_state` shouldn't be in a reserved-key filter: it's a param we're currently missing 
(causing a 400 against one issuer), so blocking it would cement that gap rather than fix it.

Kotlin has the same behaviour, so if we harden this it should land in both SDKs together, with 
client auth at the PAR endpoint. Tracking separately.

File: Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift (hunk 9 above)
Comment (abhip2565) [reply Sources/VCIClient/authorizationCodeFlow/InteractiveAuthorization/RedirectToWeb/RedirectToWebAuthorizationMethodService.swift:39]: Clarifying the expected fallback behavior:

- Keep and parse `require_pushed_authorization_requests` (default is `false` when omitted).
- If it is `true`, PAR is mandatory: a missing PAR endpoint or any PAR failure must return an error. Do **not** fall back.
- If it is `false` or omitted and a PAR endpoint is advertised, attempt PAR first. If that PAR attempt fails, fall back to the normal authorization request.
- If it is `false` or omitted and no PAR endpoint is advertised, use the normal authorization request directly.

RFC 9126 §5 saying that the endpoint's presence is sufficient means the client may use PAR; it does not mean PAR is mandatory. The `require_pushed_authorization_requests` flag is what determines whether normal authorization is allowed.

The current implementation does not meet this behavior: it removed the requirement flag, always attempts PAR when an endpoint exists, and propagates the PAR failure without fallback. Please retain the flag and add tests for the mandatory/optional success and failure paths described above.

Spec: https://www.rfc-editor.org/rfc/rfc9126.html#section-5

**Comment by Kaushikgupta469** [comment]
@coderabbitai review latest last two commit as per this changes:
Clarifying the expected fallback behavior:

    Keep and parse require_pushed_authorization_requests (default is false when omitted).
    If it is true, PAR is mandatory: a missing PAR endpoint or any PAR failure must return an error. Do not fall back.
    If it is false or omitted and a PAR endpoint is advertised, attempt PAR first. If that PAR attempt fails, fall back to the normal authorization request.
    If it is false or omitted and no PAR endpoint is advertised, use the normal authorization request directly.

RFC 9126 §5 saying that the endpoint's presence is sufficient means the client may use PAR; it does not mean PAR is mandatory. The require_pushed_authorization_requests flag is what determines whether normal authorization is allowed.

The current implementation does not meet this behavior: it removed the requirement flag, always attempts PAR when an endpoint exists, and propagates the PAR failure without fallback. Please retain the flag and add tests for the mandatory/optional success and failure paths described above.

Spec: https://www.rfc-editor.org/rfc/rfc9126.html#section-5

File: .github/workflows/push-trigger.yml (hunk 4 above)
Comment (abhip2565) [reply .github/workflows/push-trigger.yml]: @Kaushikgupta469

File: Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift (hunk 5 above)
Comment (abhip2565) [reply Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift]: @Kaushikgupta469

**Review by abhip2565 — CHANGES_REQUESTED** [review]

File: Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift (hunk 5 above)
Comment (Kaushikgupta469) [reply Sources/VCIClient/authorizationCodeFlow/AuthorizationCodeFlowService.swift]: Removed the catch block

**Review by abhip2565 — APPROVED** [review]

File: Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift [hunk 11]
```diff
@@ -5,4 +5,19 @@ struct ImplicitAuthorizationRequestData: AuthorizationRequestData {
...
+        clientMetadata: ClientMetadata,
+        pkceSession: PKCESessionManager.PKCESession,
+        scope: String,
+        pushedAuthorizationRequestEndpoint: String? = nil
+    ) {
+        self.authorizeUrl = authorizeUrl
+        self.clientMetadata = clientMetadata
+        self.pkceSession = pkceSession
+        self.scope = scope
+        self.pushedAuthorizationRequestEndpoint = pushedAuthorizationRequestEndpoint
```
[the remark this answers is not in the document]
Comment (Kaushikgupta469) [reply Sources/VCIClient/authorizationCodeFlow/ImplicitAuthorization/ImplicitAuthorizationRequestData.swift:25]: @coderabbitai already handeled Covered by the 10 PAR tests plus 3 metadata-decode tests.

File: Sources/VCIClient/authorizationServer/PushedAuthorizationRequestService.swift [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+            "redirect_uri": redirectUri,
+            "code_challenge": codeChallenge,
+            "code_challenge_method": codeChallengeMethod.rawValue,
+            "state": state,
+            "nonce": nonce,
+        ]
+
+        if let scope = scope, !scope.isEmpty {
+            params["scope"] = scope
+        }
```
[the remark this answers is not in the document]
Comment (Kaushikgupta469) [reply Sources/VCIClient/authorizationServer/PushedAuthorizationRequestService.swift:32]: @coderabbitai AuthorizationCodeFlowService passes issuerMetadata.scope ?? "default", so scope is never nil or empty.

File: Tests/VCIClientTests/authorizationServer/PushedAuthorizationRequestServiceTests.swift [hunk 13]
```diff
@@ -0,0 +1,182 @@
...
+            state: "state-123",
+            nonce: "nonce-123",
+            scope: "openid",
+            session: mock
+        )
+
+        let body = mock.capturedParams
+        XCTAssertEqual(body["scope"], "openid")
+        XCTAssertNil(body["authorization_details"])
+    }
```
[the remark this answers is not in the document]
Comment (Kaushikgupta469) [reply Tests/VCIClientTests/authorizationServer/PushedAuthorizationRequestServiceTests.swift:75]: @coderabbitai can't test a parameter that isn't supported. The current assertion is
   accurate and documents the scope-only design.

File: Sources/VCIClient/authorizationServer/PushedAuthorizationRequestService.swift [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+            )
+        }
+
+        guard let parResponse = parResponse, !parResponse.requestUri.isEmpty else {
+            throw PushedAuthorizationRequestException(
+                "Invalid PAR response from \(parEndpoint): missing request_uri"
+            )
+        }
+
+        return parResponse
```
[the remark this answers is not in the document]
Comment (Kaushikgupta469) [reply Sources/VCIClient/authorizationServer/PushedAuthorizationRequestService.swift:84]: @coderabbitai Skipping — `expires_in` is spec-required, but we never read it; only `request_uri` is used.
   Rejecting a response with a usable `request_uri` because the TTL is missing would fail a flow
   that otherwise succeeds, with no correctness gain.

File: Sources/VCIClient/exceptions/PushedAuthorizationRequestException.swift [hunk 15]
```diff
@@ -0,0 +1,25 @@
+import Foundation
+
+class PushedAuthorizationRequestException: VCIClientException {
+    init(_ message: String?) {
+        super.init(
+            code: "VCI-PAR",
```
Comment (swatigoel) [inline Sources/VCIClient/exceptions/PushedAuthorizationRequestException.swift]: this should be VCI-012 right?

File: Sources/VCIClient/exceptions/PushedAuthorizationRequestException.swift [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+    }
+
+    init(
+        message: String?,
+        issuerErrorCode: String? = nil,
+        issuerErrorDescription: String? = nil,
+        cause: Error? = nil
+    ) {
+        super.init(
+            code: "VCI-PAR",
```
Comment (swatigoel) [inline Sources/VCIClient/exceptions/PushedAuthorizationRequestException.swift]: this should be VCI-012 right?

**Review by swatigoel — CHANGES_REQUESTED** [review]

**Review by swatigoel — APPROVED** [review]
